File: LeetCode/Python/19.remove-nth-node-from-end-of-list.py
#
# @lc app=leetcode id=19 lang=python3
#
# [19] Remove Nth Node From End of List
#
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # Set the follower 1 node behind so that it is exactly one behind the node to remove  
        dummy = follower = ListNode()
        dummy.next = head
        leader = head
        
        while n > 0:
            n -= 1
            leader = leader.next

        while leader is not None:
            leader = leader.next
            follower = follower.next

        follower.next = follower.next.next

        return dummy.next


# @lc code=end

File: LeetCode/CPP/19.remove-nth-node-from-end-of-list.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=19 lang=cpp
 *
 * [19] Remove Nth Node From End of List
 */
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        ListNode *dummyNode = new ListNode();
        dummyNode->next = head;
        ListNode *fastNode = dummyNode;
        ListNode *slowNode = dummyNode;

        for (int i = 0; i < n; i++)
        {
            fastNode = fastNode->next;
        }
        while (fastNode->next != nullptr)
        {
            slowNode = slowNode->next;
            fastNode = fastNode->next;
        }
        slowNode->next = slowNode->next->next;
        return dummyNode->next;
    }
};
// @lc code=end
